nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/PutMongoRecord.java

```diff
@@ -0,0 +1,139 @@
+@InputRequirement(InputRequirement.Requirement.INPUT_REQUIRED)
+@CapabilityDescription("Bulk ingest documents into MonogDB using a configured record reader.")
+public class PutMongoRecord extends AbstractMongoProcessor {
+    static final Relationship REL_SUCCESS = new Relationship.Builder().name("success")
+            .description("All FlowFiles that are written to MongoDB are routed to this relationship").build();
+    static final Relationship REL_FAILURE = new Relationship.Builder().name("failure")
+            .description("All FlowFiles that cannot be written to MongoDB are routed to this relationship").build();
+
+    static final PropertyDescriptor RECORD_READER_FACTORY = new PropertyDescriptor.Builder()
+            .name("record-reader")
+            .displayName("Record Reader")
+            .description("Specifies the Controller Service to use for parsing incoming data and determining the data's schema")
+            .identifiesControllerService(RecordReaderFactory.class)
+            .required(true)
+            .build();
+
+    private final static Set<Relationship> relationships;
+    private final static List<PropertyDescriptor> propertyDescriptors;
+
+    static {
+        List<PropertyDescriptor> _propertyDescriptors = new ArrayList<>();
+        _propertyDescriptors.addAll(descriptors);
+        _propertyDescriptors.add(WRITE_CONCERN);
+        _propertyDescriptors.add(RECORD_READER_FACTORY);
+        propertyDescriptors = Collections.unmodifiableList(_propertyDescriptors);
+
+        final Set<Relationship> _relationships = new HashSet<>();
+        _relationships.add(REL_SUCCESS);
+        _relationships.add(REL_FAILURE);
+        relationships = Collections.unmodifiableSet(_relationships);
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    @Override
+    public List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return propertyDescriptors;
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        final FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        final RecordReaderFactory recordParserFactory = context.getProperty(RECORD_READER_FACTORY)
+                .asControllerService(RecordReaderFactory.class);
+
+        List<Document> inserts = new ArrayList<>();
+        try {
+            RecordReader reader = recordParserFactory.createRecordReader(flowFile, session.read(flowFile), getLogger());
+            RecordSchema schema = reader.getSchema();
+            Record record = null;
+            while ((record = reader.nextRecord()) != null) {
+                Document document = new Document();
+                for (String name : schema.getFieldNames()) {
+                    document.put(name, record.getValue(name));
+                }
+                inserts.add(document);
```

Comment:
I'm a little concerned about this approach. We are buffering all of this data up in the Java heap. If someone sends in a 4 GB CSV file, for instance, they are very likely to run out of heap space. I would recommend adding a "Batch Size" property, and buffering only up to the configured batch size, then sending the documents and continuing on with the next batch. While this approach has potential problems as well (you could send some of the records then route to failure), it's generally better than using up all of the JVM heap and the user has control over how much heap they would like to use to combat this possibility by tuning the Batch Size.